### PR TITLE
move implicit `Sized` predicate to end of list

### LINKED ISSUE #85998: E0277 wrong span in diagnostic with explicit `Sized` bound
Given the following code: [(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c2417382abf18396f2d8710cb3468701)

```rust
struct A<T: Sized>(T);
struct B(A<[u8]>);
```

The current output is:

```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
 --> src/lib.rs:2:10
  |
1 | struct A<T: Sized>(T);
  |          - required by this bound in `A`
2 | struct B(A<[u8]>);
  |          ^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `[u8]`
```

Ideally the output should look like:

```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
 --> src/lib.rs:2:10
  |
1 | struct A<T: Sized>(T);
  |             ----- required by this bound in `A`
2 | struct B(A<[u8]>);
  |          ^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `[u8]`
```

When the implicit `Sized` bound is added for generic type parameters, it goes at the front of the emitted predicate list. This means that its span, which is limited to the identifier of the generic, is what shows up in the error. It would be better if the error span pointed to the actual occurrence of the explicit `Sized` bound. Moving that implicit predicate to the end of the list would improve at least some of the error spans where there is an explicit `Sized` bound.

I'm working on a pull request for this issue.

@rustbot claim
@rustbot label +A-traits +A-typesystem +D-papercut

### PR DESCRIPTION
**Pull request by tlyu**
In `Bounds::predicates()`, move the implicit `Sized` predicate to the
end of the generated list. This means that if there is an explicit
`Sized` bound, it will be checked first, and any resulting
diagnostics will have a more useful span.

Fixes #85998, at least partially. ~~Based on #85979, but only the last 2 commits are new for this pull request.~~ (edit: rebased) A full fix would need to deal with where-clauses, and that seems difficult. Basically, predicates are being collected in multiple stages, and there are two places where implicit `Sized` predicates can be inserted: once for generic parameters, and once for where-clauses. I think this insertion is happening too early, and we should actually do it only at points where we collect all of the relevant trait bounds for a type parameter.

I could use some help interpreting the changes to the stderr output. It looks like reordering the predicates changed some diagnostics that don't obviously have anything to do with `Sized` bounds. Possibly some error reporting code is making assumptions about ordering of predicates? The diagnostics for src/test/ui/derives/derives-span-Hash-*.rs seem to have improved, no longer pointing at the type parameter identifier, but src/test/ui/type-alias-impl-trait/generic_duplicate_param_use9.rs became less verbose for some reason.

I also ran into an instance of #84970 while working on this, but I kind of expected that could happen, because I'm reordering predicates. I can open a separate issue on that if it would be helpful.

@estebank this seems likely to conflict (slightly?) with your work on #85947; how would you like to resolve that?

### CODE REVIEW

File: src/test/ui/issues/issue-16966.stderr [hunk 1]
```diff
@@ -1,11 +1,21 @@
...
    |
 LL |     panic!(std::default::Default::default());
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `M` declared on the function `begin_panic`
+   | 
+  ::: $SRC_DIR/std/src/panicking.rs:LL:COL
    |
+LL | pub fn begin_panic<M: Any + Send>(msg: M) -> ! {
+   |                       --- required by this bound in `begin_panic`
+   |
+   = note: cannot satisfy `_: Any`
```
Comment (estebank) [inline src/test/ui/issues/issue-16966.stderr:12]: This is arguably a regression.

**Comment by estebank** [comment]
> A full fix would need to deal with where-clauses, and that seems difficult. Basically, predicates are being collected in multiple stages, and there are two places where implicit Sized predicates can be inserted: once for generic parameters, and once for where-clauses. I think this insertion is happening too early, and we should actually do it only at points where we collect all of the relevant trait bounds for a type parameter.

I made these two changes in order to have the right spans available. I'm not sure if these changes would be enough to address what you're mentioning here.

https://github.com/rust-lang/rust/pull/85947/commits/29c8a9f4cb31cc9dcd1feec2c8eee90eb1d4e9e1#diff-0a61b538a3cec072c76fecae4635af6a12ec3256860029ac70549c2aa53ab394R1480-R1485

https://github.com/rust-lang/rust/pull/85947/commits/29c8a9f4cb31cc9dcd1feec2c8eee90eb1d4e9e1#diff-ac1e6e554bac799e1e2d3886bea8fdc0d0bf04b20d33df59f5a8e92c41987ac6R2137-R2139

**Comment by estebank** [comment]
r? @estebank 
r=me after potentially addressing the src/test/ui/issues/issue-16966.stderr case, but if you find it difficult to do (I would just remove the note in every case where it is `_: Something`) or lack the time, we can land as is.

**Comment by tlyu** [comment]
Thanks for the review! I'll be taking a closer look at the things you pointed out and will probably reply in more detail soon.

>>I also ran into an instance of #84970 while working on this, but I kind of expected that could happen, because I'm reordering predicates. I can open a separate issue on that if it would be helpful.

>Did this happen on tests or while working on rustc? Did it happen with --keep-stage 1? If the problem is one of "we might hit this in CI" it's less concerning to me than "users will hit a new fingerprint issue in stable in 12 weeks".

I opened issue #86013 with more details. Basically, when I either applied or reverted the patches in this PR, the fingerprints changed and I got the ICE on a specific UI test case. For what it's worth, I built the stage1 compiler and libstd using `./x.py build -i`, and did not encounter ICEs while building the compiler itself.

As for whether this will cause fingerprint errors when this lands in stable? I'm not sure I know enough about the details of incremental compilation. Do we have some way of invalidating incremental compilation caches when there are internal changes to the compiler? (I hope so.) If so, how is this done? Is it part of the release process? Automatically according to the commit-hash stored during build?

**Comment by tlyu** [comment]
>src/test/ui/type-alias-impl-trait/generic_duplicate_param_use9.rs seemed to not get more verbose, things just moved around, which make sense.

if you compare
https://github.com/rust-lang/rust/pull/86011/files#diff-06109f234a5a60a647998669551ba82a0a549f34b4741bd5da4a449afbbe14a6L13
with
https://github.com/rust-lang/rust/pull/86011/files#diff-06109f234a5a60a647998669551ba82a0a549f34b4741bd5da4a449afbbe14a6R37
you'll notice `(A, B, <A as Foo>::Bar)` got deleted from the error message.

For issue-16966, I found this area to be somewhat suspicious:
https://github.com/rust-lang/rust/blob/6c2dd251bbff03c7a3092d43fb5b637eca0810e3/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs#L1539-L1543
and I'm attempting to investigate. It maybe that the type inference ambiguity causes a bunch of simultaneous errors and we previously happened to get the `Sized` trait predicate first, outputting the more terse error E0282 instead of E0283.

**Comment by tlyu** [comment]
What I thought was the "obvious" approach to quieting the new E0283 in issue-16966 has some nontrivial fallout. Maybe the smallest change to the diagnostics is what we already have in this pull request? Do you think we need to make additional changes in this pull request?

I tried to see what would happen if I made this block in `maybe_report_ambiguity()` either always or never execute:

https://github.com/rust-lang/rust/blob/6c2dd251bbff03c7a3092d43fb5b637eca0810e3/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs#L1539-L1543

I wound up with these outputs from the UI tests:
always:
[always-E0282.txt](https://github.com/rust-lang/rust/files/6613114/always-E0282.txt)
never:
[never-E0282.txt](https://github.com/rust-lang/rust/files/6613115/never-E0282.txt)

The "never" approach generates a lot more diffs in the stderr output than the "always" approach. (51 failures with "never", vs 19 failures with "always")

A quick glance at the output from the "always" approach shows that the diffs are largely making some vague verbose diagnostics less verbose.

My guess (partially confirmed by disabling some of the verbosity suppression) is that in the case of issue-16966, the type inference ambiguity was causing every trait bound to not be satisfied, and prior to my changes, we were luckily hitting `Sized` first and ending up with E0282, but are now getting E0283 because the first unsatisfied bound is no longer `Sized`.

Any opinions on whether the deleted verbosity with "always" was at all helpful? It sounds like you might be concerned that the E0283 now appearing with issue-16966 is too verbose? If so, how do you feel about the existing instances of E0283 verbosity that are deleted in the "always" approach?

**Comment by estebank** [comment]
At a very quick glance, it seems like the always case removes a bunch of useful suggestions, while the never case is just more verbose and perhaps slightly misleading...

I'll try to take a look at this PR with fresh eyes tomorrow, but I think we might want to merge as is.

**Comment by tlyu** [comment]
@estebank
> Did this happen on tests or while working on rustc? Did it happen with --keep-stage 1? If the problem is one of "we might hit this in CI" it's less concerning to me than "users will hit a new fingerprint issue in stable in 12 weeks".

I wrote an update in https://github.com/rust-lang/rust/issues/86013#issuecomment-860252506

It seems unlikely to affect users of released versions because invalidation of the incremental caches seems to honor changes in the commit hash.

**Comment by tlyu** [comment]
Opened #86371 with some discouraging findings. Maybe the better approach is to detect whether we are dealing with violating an implicit `Sized` bound, and adjust the span to point at any explicit `Sized` bound that happens to be present, instead of to the type parameter name.

**Comment by estebank** [comment]
@tlyu can you rebase? After that r=me, but I would still like to somehow fix `src/test/ui/issues/issue-16966.stderr`.

**Comment by tlyu** [comment]
@estebank Rebase done. I looked at #16966 and the change in the stderr in this PR isn't a necessarily a regression of the substance of that issue, which is about duplication of error messages. (The change in `src/test/ui/issues/issue-16966.stderr` is more (potentially-useless) verbosity than there used to be, but there's no repetition.)

I think there's a big rabbit hole here of various error reporting heuristics that use the failure to satisfy the implicit `Sized` constraint as a way of suppressing excess verbosity. I'm not sure what to do about them, because in theory, the ordering of trait bounds should be irrelevant to type checking results, right?

**Comment by tlyu** [comment]
Rebased again.

**Comment by camelid** [comment]
triage: @tlyu could you rebase again? Then perhaps ping estebank to let them know the PR is ready to be approved. Thanks!

**Comment by JohnCSimon** [comment]
triage this seems to be something that changes often

@tlyu could you rebase again? 

and then set the PR to S-waiting-on-review

**Comment by JohnCSimon** [comment]
@tlyu could you rebase again?

**Comment by estebank** [comment]
Looking at the output after the rebase, I am much more comfortable landing this as is. Apologies for the delay in landing this!

@bors r+

**Pull request merged**